aptos-move/replay-benchmark/src/generator.rs

```diff
@@ -1,105 +1,166 @@
+pub struct InputOutputDiffGenerator {
     debugger: AptosDebugger,
-    begin_version: Version,
-    end_version: Version,
     override_config: OverrideConfig,
+    allow_different_gas_usage: bool,
 }
 
-impl BenchmarkGenerator {
-    /// Generates a sequence of [Block] for benchmarking.
-    pub async fn generate_blocks(
+impl InputOutputDiffGenerator {
+    /// Generates a sequence of inputs (pre-block states) for benchmarking. Also, returns a vector
+    /// of output diffs when transactions are executed with these inputs (recall that inputs may be
+    /// different from on-chain due to overrides).
+    pub(crate) async fn generate(
         debugger: AptosDebugger,
-        begin_version: Version,
-        end_version: Version,
         override_config: OverrideConfig,
-    ) -> anyhow::Result<Vec<Block>> {
+        txn_blocks: Vec<TransactionBlock>,
+        allow_different_gas_usage: bool,
+    ) -> anyhow::Result<(Vec<ReadSet>, Vec<Vec<TransactionDiff>>)> {
         let generator = Arc::new(Self {
             debugger,
-            begin_version,
-            end_version,
             override_config,
+            allow_different_gas_usage,
         });
 
-        let limit = generator.end_version - generator.begin_version + 1;
-        let (txns, _) = generator
-            .debugger
-            .get_committed_transactions(generator.begin_version, limit)
-            .await?;
-        let txn_blocks = generator.partition(txns);
-
         let num_generated = Arc::new(AtomicU64::new(0));
         let num_blocks = txn_blocks.len();
 
         let mut tasks = Vec::with_capacity(num_blocks);
-        for (begin, txn_block) in txn_blocks {
+        for txn_block in txn_blocks {
             let task = tokio::task::spawn_blocking({
                 let generator = generator.clone();
                 let num_generated = num_generated.clone();
                 move || {
                     let start_time = Instant::now();
-                    let block = generator.generate_block(begin, txn_block);
+                    let (inputs, diffs) = generator.generate_inputs_with_diffs(txn_block);
                     let time = start_time.elapsed().as_secs();
                     println!(
-                        "Generated block {}/{} in {}s",
+                        "Generated inputs and computed diffs for block {}/{} in {}s",
                         num_generated.fetch_add(1, Ordering::SeqCst) + 1,
                         num_blocks,
                         time
                     );
-                    block
+                    (inputs, diffs)
                 }
             });
             tasks.push(task);
         }
 
-        let mut blocks = Vec::with_capacity(tasks.len());
+        let mut all_inputs = Vec::with_capacity(tasks.len());
+        let mut all_diffs = Vec::with_capacity(tasks.len());
         for task in tasks {
-            blocks.push(task.await?);
+            let (inputs, diffs) = task.await?;
```

Comment:
(unless I'm missing smt) since we don't need generate to be async, and we aren't joining any tasks, I think we can block_on here and make generate not be async.

Reply:
But how do you block on? The runtime is created in main, so down here there is no context? Although we can move it creation to here, but so far all CLI examples I have seen in the codebase were using this pattern

Reply:
block_on(task)? unless I'm missing smt

Reply:
I meant that here we do not have access to tokio's runtime, it is only created in main and unless we pass it to here you cannot really call `block_on`? i.e.
```rust
let runtime  = Runtime::new().unwrap();
runtime.block_on(async { /*task here*/ })
```
I found that you can use `Handle::current()` to get the current runtime, not sure what is the best practice here, wdyt?

Reply:
Btw, I will merge as is, and related async comments we can land/address as a separate PR